contracts/products/SoteriaCoverageProduct.sol

```diff
@@ -0,0 +1,503 @@
+        uint256 policyID = policyByOwner(msg.sender);
+        require(policyID > 0, "invalid policy");
+        uint256 currentCoverAmount = coverAmountOf(policyID);
+        require(assessRisk(currentCoverAmount, newCoverAmount_), "cannot accept that risk");
+        uint256 coverAmount = activeCoverAmount();
+        coverAmount = coverAmount + newCoverAmount_ - currentCoverAmount;
+        _coverAmountOf[policyID] = newCoverAmount_;
+        _activeCoverAmount = coverAmount;
+        _updatePolicyManager(coverAmount);
+        emit PolicyUpdated(policyID);
+    }
+
+    /**
+     * @notice Deposits funds for policy holders.
+     * @param policyholder_ The holder of the policy.
+    */
+    function deposit(address policyholder_) external payable override whileUnpaused {
+        _deposit(policyholder_, msg.value);
+    }
+
+    /**
+     * @notice Charge premiums for each policy holder.
+     * @param holders_ The policy holders.
+     * @param premiums_ The premium amounts in `wei` per policy holder.
+    */
+    function chargePremiums(address[] calldata holders_, uint256[] calldata premiums_) external payable override onlyGovernance whileUnpaused {
+        uint256 count = holders_.length;
+        require(count == premiums_.length, "length mismatch");
+        require(count <= policyCount(), "policy count exceeded");
+        uint256 amountToPayTreasury = 0;
+
+        for (uint256 i = 0; i < count; i++) {
+            // skip computation if policy inactive (coverAmount == 0)
+            if (coverAmountOf(policyByOwner(holders_[i])) == 0) {
+                continue;
+            }
+
+            if (_funds[holders_[i]] >= premiums_[i]) {
+                amountToPayTreasury += premiums_[i];
+                _funds[holders_[i]] -= premiums_[i];
+                emit PremiumCharged(holders_[i], premiums_[i]);
+            } else {
+                uint256 partialPremium = _funds[holders_[i]];
+                amountToPayTreasury += partialPremium;
+                _funds[holders_[i]] = 0;
+                // turn off the policy
+                _closePolicy(holders_[i]);
+                emit PremiumPartiallyCharged(holders_[i], premiums_[i], partialPremium);
+            }  
+        }
+        // transfer premium to the treasury
+        ITreasury(payable(_registry.treasury())).routePremiums{value: amountToPayTreasury}();
+    }
+
+    /**
+     * @notice Cancel and burn a policy.
+     * User will receive their deposited funds.
+     * Can only be called by the policyholder.
+     * @param policyID_ The ID of the policy.
+     */
+     function cancelPolicy(uint256 policyID_) external override nonReentrant {
+        require(_exists(policyID_), "invalid policy");
+        require(ownerOfPolicy(policyID_) == msg.sender, "!policyholder");
+        uint256 refundAmount = funds(msg.sender);
+        _activeCoverAmount -= coverAmountOf(policyID_);
+        _coverAmountOf[policyID_] = 0;
+
```

Comment:
Here, should we add `_totalPolicyCount =- 1;`?

Reply:
SoteriaProduct is abstracted away from the PolicyManager. The policyId from SoteriaProduct differs from the PolicyManager because Soteria has its own policy related logics. (You can think that Soteria is sort of like PolicyManager).

We don't burn policy for Soteria. After the cancellation user can activate his policy with same policy id.

Reply:
My bad, didn't realise that this mechanism was specified in the specs - https://solacefi.atlassian.net/browse/SF-4?focusedCommentId=10001